### PR TITLE
[ecash-lib] Add support for LEGACY sighash

### PR DESCRIPTION
**Pull request by EyeOfPython**
This is the OG sighash from Bitcoin, which is invalid on eCash, but still can be used on Bitcoin, Dogecoin and others.

Note: This cannot be tested on eCash as regtest will never run legacy sigchecks. There, IS_LEGACY_MODE must be false.

### CODE REVIEW

File: modules/ecash-lib/tests/txBuilder.test.ts [hunk 1]
```diff
@@ -471,16 +518,20 @@ describe('TxBuilder', () => {
...
         spendTx = txBuild.sign(ecc, 100000, 546);
         await chronik.broadcastTx(spendTx.ser());
+        estimatedSize = txBuild.sign(new EccDummy(), 100000, 546).serSize();
         expect(spendTx.outputs[1].value).to.equal(
-            BigInt(40000 - 100 * spendTx.serSize()),
+            BigInt(40000 - 100 * estimatedSize),
         );
 
-        // 120ksats/kB, deletes leftover output
+        // 119ksats/kB, deletes leftover output
```
Comment (EyeOfPython) [inline modules/ecash-lib/tests/txBuilder.test.ts:526]:
```suggestion
        // 117.6ksats/kB, deletes leftover output
```

File: modules/ecash-lib/src/unsignedTx.ts [hunk 2]
```diff
@@ -119,6 +219,29 @@ export class UnsignedTxInput {
                 ? redeemScript
                 : redeemScript.cutOutCodesep(nCodesep);
 
+        // Sign LEGACY signatures that don't use SIGHASH_FORKID
+        if (sigHashType.variant == SigHashTypeVariant.LEGACY) {
```
Comment (BytesOfMan) [inline modules/ecash-lib/src/unsignedTx.ts:223]: any reason for not `sigHashType.variant === SigHashTypeVariant.LEGACY)` ?

`===` vs `==`

File: modules/ecash-lib/src/unsignedTx.ts [hunk 3]
```diff
@@ -83,6 +90,102 @@ export interface SighashPreimage {
...
+    writer: Writer,
+    tx: Tx,
+    scriptCode: Script,
+    inputIdx: number,
+    sigHashType: SigHashType,
+) {
+    const hasAnyoneCanPay =
+        sigHashType.inputType == SigHashTypeInputs.ANYONECANPAY;
+
+    const writeLegacyScriptCode = () => {
```
Comment (BytesOfMan) [inline modules/ecash-lib/src/unsignedTx.ts:105]: any advantage of keeping these function definitions outside of `writeLegacyPreImage`?

mb not used anywhere else but I'm not familiar with the rationale for keeping them all in this function, vs defining them outside then using them here.

File: modules/ecash-lib/src/unsignedTx.ts [hunk 4]
```diff
@@ -83,6 +90,102 @@ export interface SighashPreimage {
...
+        // Filter out all code separators
+        while ((nextOp = ops.next()) !== undefined) {
+            if (isPushOp(nextOp) || nextOp != OP_CODESEPARATOR) {
+                newOps.push(nextOp);
+            }
+        }
+        Script.fromOps(newOps).writeWithSize(writer);
+    };
+
+    const writeLegacyInput = (idx: number) => {
```
Comment (BytesOfMan) [inline modules/ecash-lib/src/unsignedTx.ts:118]: same

File: modules/ecash-lib/src/unsignedTx.ts [hunk 5]
```diff
@@ -83,6 +90,102 @@ export interface SighashPreimage {
...
+                sigHashType.outputType == SigHashTypeOutputs.NONE)
+        ) {
+            // let the others update at will
+            writer.putU32(0);
+        } else {
+            writer.putU32(input.sequence ?? DEFAULT_SEQUENCE);
+        }
+    };
+
+    const writeLegacyOutput = (idx: number) => {
```
Comment (BytesOfMan) [inline modules/ecash-lib/src/unsignedTx.ts:147]: same

File: modules/ecash-lib/src/unsignedTx.ts (hunk 3 above)
Comment (EyeOfPython) [reply modules/ecash-lib/src/unsignedTx.ts:105]: The reason is that these are highly specific functions that really should never be used outside.

It also mimics the C++ code a little where these are encapsulated all under one class.

File: modules/ecash-lib/src/unsignedTx.ts (hunk 2 above)
Comment (EyeOfPython) [reply modules/ecash-lib/src/unsignedTx.ts:223]: Good point.

**Review by BytesOfMan — APPROVED** [review]

**Pull request merged**